server/src/main/java/org/elasticsearch/action/SingleResultDeduplicator.java

```diff
@@ -59,10 +61,10 @@ public void execute(ActionListener<T> listener) {
                 return;
             }
         }
-        doExecute(listener);
+        doExecute(threadContext.newStoredContext(), ContextPreservingActionListener.wrapPreservingContext(listener, threadContext));
```

Comment:
I think a more natural API would be to take a `Supplier<StoredContext> restorableContext` and pass in `threadContext.newRestorableContext(false)` here. That way you can wrap the inner call with `try (var ignored = restorableContext.get())` instead.

Reply:
The `storedContext` needs to be restored _before_ doing work, e.g.:
```
storedContext.restore();
doWork();
```

 while `restorableContext` is better at restore the context _after_ doing work, e.g.:
```
try (var ignore = restorableContext.get()) {
    doWork();
}
```

What I need is the former. So using `storedContext` seems fine to me.